[test] switching from nosetests to pytest?

In the process of checking the coverage #3205 for our python test using [coverage](https://coverage.readthedocs.io/), I stumbled on this in the documentation. 

> Nose has been unmaintained for a long time. You should seriously consider adopting a different test runner.

Let's see [nose documentation](https://nose.readthedocs.io/en/latest/) and [code](https://github.com/nose-devs/nose).

> Nose has been in maintenance mode for the past several years and will likely cease without a new person/team to take over maintainership. New projects should consider using Nose2, py.test, or just plain unittest/unittest2. 

last commit in March 2016, 4 years ago. So probably it's time for us to move on. Not urgent, but we can at least test what it would mean to switch to pytest.

http://doc.pytest.org/en/latest/nose.html is good to know about.

I tried running our nose tests w/ pytest but run into the following:

```
🐓 pytest tests/
=================================================================================== test session starts ====================================================================================
platform darwin -- Python 3.7.4, pytest-5.3.5, py-1.8.1, pluggy-0.13.1
rootdir: /Users/miket/dev/webcompat.com
plugins: mozlog-5.0
collected 0 items / 13 errors                                                                                                                                                              

========================================================================================== ERRORS ==========================================================================================
_______________________________________________________________________ ERROR collecting tests/unit/test_api_urls.py _______________________________________________________________________
ImportError while importing test module '/Users/miket/dev/webcompat.com/tests/unit/test_api_urls.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/test_api_urls.py:17: in <module>
    import webcompat
E   ModuleNotFoundError: No module named 'webcompat'
________________________________________________________________________ ERROR collecting tests/unit/test_config.py ________________________________________________________________________
ImportError while importing test module '/Users/miket/dev/webcompat.com/tests/unit/test_config.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/test_config.py:13: in <module>
    from config import update_status_config
E   ModuleNotFoundError: No module named 'config'
_____________________________________________________________________ ERROR collecting tests/unit/test_console_logs.py _____________________________________________________________________
ImportError while importing test module '/Users/miket/dev/webcompat.com/tests/unit/test_console_logs.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/test_console_logs.py:18: in <module>
    from webcompat import app  # noqa
E   ModuleNotFoundError: No module named 'webcompat'
_________________________________________________________________________ ERROR collecting tests/unit/test_form.py _________________________________________________________________________
ImportError while importing test module '/Users/miket/dev/webcompat.com/tests/unit/test_form.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/test_form.py:11: in <module>
    import webcompat
E   ModuleNotFoundError: No module named 'webcompat'
_______________________________________________________________________ ERROR collecting tests/unit/test_helpers.py ________________________________________________________________________
ImportError while importing test module '/Users/miket/dev/webcompat.com/tests/unit/test_helpers.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/test_helpers.py:18: in <module>
    import webcompat
E   ModuleNotFoundError: No module named 'webcompat'
_____________________________________________________________________ ERROR collecting tests/unit/test_http_caching.py _____________________________________________________________________
ImportError while importing test module '/Users/miket/dev/webcompat.com/tests/unit/test_http_caching.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/test_http_caching.py:15: in <module>
    import webcompat  # noqa
E   ModuleNotFoundError: No module named 'webcompat'
________________________________________________________________________ ERROR collecting tests/unit/test_issues.py ________________________________________________________________________
ImportError while importing test module '/Users/miket/dev/webcompat.com/tests/unit/test_issues.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/test_issues.py:18: in <module>
    from webcompat.issues import report_issue
E   ModuleNotFoundError: No module named 'webcompat'
______________________________________________________________________ ERROR collecting tests/unit/test_rendering.py _______________________________________________________________________
ImportError while importing test module '/Users/miket/dev/webcompat.com/tests/unit/test_rendering.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/test_rendering.py:15: in <module>
    import webcompat  # noqa
E   ModuleNotFoundError: No module named 'webcompat'
___________________________________________________________________ ERROR collecting tests/unit/test_tools_changelog.py ____________________________________________________________________
ImportError while importing test module '/Users/miket/dev/webcompat.com/tests/unit/test_tools_changelog.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/test_tools_changelog.py:11: in <module>
    from tools import changelog
E   ModuleNotFoundError: No module named 'tools'
_______________________________________________________________________ ERROR collecting tests/unit/test_topsites.py _______________________________________________________________________
ImportError while importing test module '/Users/miket/dev/webcompat.com/tests/unit/test_topsites.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/test_topsites.py:14: in <module>
    from tools import topsites
E   ModuleNotFoundError: No module named 'tools'
_______________________________________________________________________ ERROR collecting tests/unit/test_uploads.py ________________________________________________________________________
ImportError while importing test module '/Users/miket/dev/webcompat.com/tests/unit/test_uploads.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/test_uploads.py:23: in <module>
    from webcompat import app  # noqa
E   ModuleNotFoundError: No module named 'webcompat'
_________________________________________________________________________ ERROR collecting tests/unit/test_urls.py _________________________________________________________________________
ImportError while importing test module '/Users/miket/dev/webcompat.com/tests/unit/test_urls.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/test_urls.py:18: in <module>
    import webcompat  # noqa
E   ModuleNotFoundError: No module named 'webcompat'
_______________________________________________________________________ ERROR collecting tests/unit/test_webhook.py ________________________________________________________________________
ImportError while importing test module '/Users/miket/dev/webcompat.com/tests/unit/test_webhook.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unit/test_webhook.py:17: in <module>
    import webcompat
E   ModuleNotFoundError: No module named 'webcompat'
===================================================================================== warnings summary =====================================================================================
env/lib/python3.7/site-packages/jinja2/utils.py:485
  /Users/miket/dev/webcompat.com/env/lib/python3.7/site-packages/jinja2/utils.py:485: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import MutableMapping

env/lib/python3.7/site-packages/jinja2/runtime.py:318
  /Users/miket/dev/webcompat.com/env/lib/python3.7/site-packages/jinja2/runtime.py:318: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Mapping

-- Docs: https://docs.pytest.org/en/latest/warnings.html
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 13 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================== 2 warnings, 13 errors in 0.39s ==============================================================================
(
```

Probably some trick to fix the imports.

Maybe if we can fix this, we just start running tests via `pytest`, and then slowly port existing stuff to use `pytest` idioms.

This is a better documentation
https://docs.pytest.org/en/latest/goodpractices.html

In `webcompat.com/setup.py`

```python
from setuptools import setup, find_packages

setup(name="WEBCOMPAT", packages=find_packages())
```

and then 

```bash
pip install -e .
```

we can run the tests with `pytest -x tests`

It will stop at the first failure because of `-x`. we get

```
================================================================= test session starts =================================================================
platform darwin -- Python 3.7.4, pytest-5.3.5, py-1.8.1, pluggy-0.13.1
rootdir: /Users/karl/code/webcompat.com
collected 152 items                                                                                                                                   

tests/unit/test_api_urls.py ...F

====================================================================== FAILURES =======================================================================
__________________________________________________________ TestAPIURLs.test_api_patch_issue ___________________________________________________________

thing = <flask.blueprints.Blueprint object at 0x112981290>, comp = 'endpoints', import_path = 'webcompat.api.endpoints'

    def _dot_lookup(thing, comp, import_path):
        try:
>           return getattr(thing, comp)
E           AttributeError: 'Blueprint' object has no attribute 'endpoints'

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/mock.py:1101: AttributeError

During handling of the above exception, another exception occurred:

self = <unit.test_api_urls.TestAPIURLs testMethod=test_api_patch_issue>

    def test_api_patch_issue(self):
        """Patching the issue is working only with certain circumstances."""
        with webcompat.app.app_context():
            webcompat.app.config.update(STATUSES=STATUSES)
            # Incompatible state and status
            data = {'state': 'closed', 'milestone': 2}
            patch_data = json.dumps(data)
            rv = self.app.patch('/api/issues/1/edit', data=patch_data,
                                environ_base=headers)
            self.assertEqual(rv.status_code, 403)
            # Too many elements in the JSON
            data = {'state': 'open', 'milestone': 2, 'foobar': 'z'}
            patch_data = json.dumps(data)
            rv = self.app.patch('/api/issues/1/edit', data=patch_data,
                                environ_base=headers)
            self.assertEqual(rv.status_code, 403)
            # Valid request
>           with patch('webcompat.api.endpoints.proxy_request') as github_data:

tests/unit/test_api_urls.py:175: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/mock.py:1257: in __enter__
    self.target = self.getter()
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/mock.py:1427: in <lambda>
    getter = lambda: _importer(target)
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/mock.py:1114: in _importer
    thing = _dot_lookup(thing, comp, import_path)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

thing = <flask.blueprints.Blueprint object at 0x112981290>, comp = 'endpoints', import_path = 'webcompat.api.endpoints'

    def _dot_lookup(thing, comp, import_path):
        try:
            return getattr(thing, comp)
        except AttributeError:
            __import__(import_path)
>           return getattr(thing, comp)
E           AttributeError: 'Blueprint' object has no attribute 'endpoints'

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/mock.py:1104: AttributeError
================================================================== warnings summary ===================================================================
tests/unit/test_uploads.py:45
  /Users/karl/code/webcompat.com/tests/unit/test_uploads.py:45: PytestCollectionWarning: cannot collect test class 'TestingFileStorage' because it has a __init__ constructor (from: tests/unit/test_uploads.py)
    class TestingFileStorage(FileStorage):

-- Docs: https://docs.pytest.org/en/latest/warnings.html
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
======================================================= 1 failed, 3 passed, 1 warning in 0.92s ========================================================
```

And the full test suite: `pytest tests`

```
================================================================= test session starts =================================================================
platform darwin -- Python 3.7.4, pytest-5.3.5, py-1.8.1, pluggy-0.13.1
rootdir: /Users/karl/code/webcompat.com
collected 152 items                                                                                                                                   

tests/unit/test_api_urls.py ...F.....                                                                                                           [  5%]
tests/unit/test_config.py ..                                                                                                                    [  7%]
tests/unit/test_console_logs.py .....                                                                                                           [ 10%]
tests/unit/test_form.py .................                                                                                                       [ 21%]
tests/unit/test_helpers.py ............................                                                                                         [ 40%]
tests/unit/test_http_caching.py ...                                                                                                             [ 42%]
tests/unit/test_issues.py ......                                                                                                                [ 46%]
tests/unit/test_rendering.py ......                                                                                                             [ 50%]
tests/unit/test_tools_changelog.py ..                                                                                                           [ 51%]
tests/unit/test_topsites.py ...                                                                                                                 [ 53%]
tests/unit/test_uploads.py ...                                                                                                                  [ 55%]
tests/unit/test_urls.py .................................                                                                                       [ 76%]
tests/unit/test_webhook.py ...................................                                                                                  [100%]

====================================================================== FAILURES =======================================================================
__________________________________________________________ TestAPIURLs.test_api_patch_issue ___________________________________________________________

thing = <flask.blueprints.Blueprint object at 0x10d5f8250>, comp = 'endpoints', import_path = 'webcompat.api.endpoints'

    def _dot_lookup(thing, comp, import_path):
        try:
>           return getattr(thing, comp)
E           AttributeError: 'Blueprint' object has no attribute 'endpoints'

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/mock.py:1101: AttributeError

During handling of the above exception, another exception occurred:

self = <unit.test_api_urls.TestAPIURLs testMethod=test_api_patch_issue>

    def test_api_patch_issue(self):
        """Patching the issue is working only with certain circumstances."""
        with webcompat.app.app_context():
            webcompat.app.config.update(STATUSES=STATUSES)
            # Incompatible state and status
            data = {'state': 'closed', 'milestone': 2}
            patch_data = json.dumps(data)
            rv = self.app.patch('/api/issues/1/edit', data=patch_data,
                                environ_base=headers)
            self.assertEqual(rv.status_code, 403)
            # Too many elements in the JSON
            data = {'state': 'open', 'milestone': 2, 'foobar': 'z'}
            patch_data = json.dumps(data)
            rv = self.app.patch('/api/issues/1/edit', data=patch_data,
                                environ_base=headers)
            self.assertEqual(rv.status_code, 403)
            # Valid request
>           with patch('webcompat.api.endpoints.proxy_request') as github_data:

tests/unit/test_api_urls.py:175: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/mock.py:1257: in __enter__
    self.target = self.getter()
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/mock.py:1427: in <lambda>
    getter = lambda: _importer(target)
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/mock.py:1114: in _importer
    thing = _dot_lookup(thing, comp, import_path)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

thing = <flask.blueprints.Blueprint object at 0x10d5f8250>, comp = 'endpoints', import_path = 'webcompat.api.endpoints'

    def _dot_lookup(thing, comp, import_path):
        try:
            return getattr(thing, comp)
        except AttributeError:
            __import__(import_path)
>           return getattr(thing, comp)
E           AttributeError: 'Blueprint' object has no attribute 'endpoints'

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/mock.py:1104: AttributeError
================================================================== warnings summary ===================================================================
tests/unit/test_uploads.py:45
  /Users/karl/code/webcompat.com/tests/unit/test_uploads.py:45: PytestCollectionWarning: cannot collect test class 'TestingFileStorage' because it has a __init__ constructor (from: tests/unit/test_uploads.py)
    class TestingFileStorage(FileStorage):

tests/unit/test_issues.py::TestIssue::test_report_issue_returns_number
  /Users/karl/code/webcompat.com/tests/unit/test_issues.py:50: DeprecationWarning: Please use assertEqual instead.
    self.assertEquals(rv.get('number'), 2)

tests/unit/test_rendering.py::TestURIContent::test_form_rendering
  /Users/karl/code/webcompat.com/env/lib/python3.7/site-packages/jinja2/lexer.py:648: DeprecationWarning: invalid escape sequence '\d'
    .decode("unicode-escape")

-- Docs: https://docs.pytest.org/en/latest/warnings.html
====================================================== 1 failed, 151 passed, 3 warnings in 1.91s ======================================================
```

so basically 1 fail and 3 warnings to fix.
And we could switch to pytest.

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -58,7 +58,7 @@ jobs:
             . env/bin/activate
             npm run lint
             npm run build
-            nosetests
+            pytest tests
             python run.py -t &
             sleep 5
             node --max-old-space-size=8192 ./tests/functional/_intern.js --reporters="runner" --firefoxBinary=`which firefox`

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -58,6 +58,7 @@ jobs:
             . env/bin/activate
             npm run lint
             npm run build
+            pip install -e .
             pytest tests
             python run.py -t &
             sleep 5

--- a/.gitignore
+++ b/.gitignore
@@ -17,6 +17,9 @@ config/secrets.py
 package-lock.json
 tmp/
 
+# Ignore the installable app.
+*.egg-info
+
 # The data folder contains information that shouldn't live in version control.
 data/*
 

--- a/config/requirements-dev.txt
+++ b/config/requirements-dev.txt
@@ -1,4 +1,4 @@
 # This is for dev purpose
 -r requirements.txt
-nose==1.3.7
 pycodestyle==2.5.0
+pytest==5.3.5

--- a/setup.py
+++ b/setup.py
@@ -1,0 +1,3 @@
+from setuptools import setup, find_packages
+
+setup(name="WEBCOMPAT", packages=find_packages())

--- a/tests/unit/test_api_urls.py
+++ b/tests/unit/test_api_urls.py
@@ -174,7 +174,7 @@ class TestAPIURLs(unittest.TestCase):
                                 environ_base=headers)
             self.assertEqual(rv.status_code, 403)
 
-    @patch('webcompat.helpers.proxy_request')
+    @patch('webcompat.api.endpoints.proxy_request')
     def test_api_patch_issue_valid_request(self, github_data):
         """Patching the issue - Valid request."""
         with webcompat.app.app_context():

--- a/tests/unit/test_api_urls.py
+++ b/tests/unit/test_api_urls.py
@@ -155,31 +155,36 @@ class TestAPIURLs(unittest.TestCase):
         self.assertEqual(rv.content_type, 'application/json')
         self.assertEqual(json_body['status'], 404)
 
-    def test_api_patch_issue(self):
-        """Patching the issue is working only with certain circumstances."""
+    def test_api_patch_issue_state_status(self):
+        """Patching the issue - Incompatible state and status."""
         with webcompat.app.app_context():
             webcompat.app.config.update(STATUSES=STATUSES)
-            # Incompatible state and status
             data = {'state': 'closed', 'milestone': 2}
             patch_data = json.dumps(data)
             rv = self.app.patch('/api/issues/1/edit', data=patch_data,
                                 environ_base=headers)
             self.assertEqual(rv.status_code, 403)
-            # Too many elements in the JSON
+
+    def test_api_patch_issue_too_many_json_elements(self):
+        """Patching the issue - Too many elements in the JSON."""
+        with webcompat.app.app_context():
             data = {'state': 'open', 'milestone': 2, 'foobar': 'z'}
             patch_data = json.dumps(data)
             rv = self.app.patch('/api/issues/1/edit', data=patch_data,
                                 environ_base=headers)
             self.assertEqual(rv.status_code, 403)
-            # Valid request
-            with patch('webcompat.api.endpoints.proxy_request') as github_data:
-                github_data.return_value = mock_api_response(
-                    {'status_code': 200, 'content': '[]'})
-                data = {'state': 'open', 'milestone': 2}
-                patch_data = json.dumps(data)
-                rv = self.app.patch('/api/issues/1/edit', data=patch_data,
-                                    environ_base=headers)
-                self.assertEqual(rv.status_code, 200)
+
+    @patch('webcompat.helpers.proxy_request')
+    def test_api_patch_issue_valid_request(self, github_data):
+        """Patching the issue - Valid request."""
+        with webcompat.app.app_context():
+            github_data.return_value = mock_api_response(
+                {'status_code': 200, 'content': '[]'})
+            data = {'state': 'open', 'milestone': 2}
+            patch_data = json.dumps(data)
+            rv = self.app.patch('/api/issues/1/edit', data=patch_data,
+                                environ_base=headers)
+            self.assertEqual(rv.status_code, 200)
 
 
 if __name__ == '__main__':

--- a/tests/unit/test_issues.py
+++ b/tests/unit/test_issues.py
@@ -47,7 +47,7 @@ class TestIssue(unittest.TestCase):
             ('url', 'http://3139.example.com'),
             ('username', ''), ])
         rv = report_issue(form, True)
-        self.assertEquals(rv.get('number'), 2)
+        self.assertEqual(rv.get('number'), 2)
 
     @patch.object(requests, 'post')
     def test_report_private_issue_returns_nothing(self, mockpost):

--- a/tests/unit/test_uploads.py
+++ b/tests/unit/test_uploads.py
@@ -42,7 +42,7 @@ def check_rv_format(self, resp):
         )
 
 
-class TestingFileStorage(FileStorage):
+class MockingFileStorage(FileStorage):
     """Test-only File Storage class.
 
     This is a helper for testing upload behavior in your application. You
@@ -121,13 +121,13 @@ class TestUploads(unittest.TestCase):
 
             # The reason why we are defining it in here and not outside
             # this method is that we are setting the filename of the
-            # TestingFileStorage to be the one in the for loop. This way
+            # MockingFileStorage to be the one in the for loop. This way
             # we can ensure that the filename that we are "uploading"
             # is the same as the one being used by the application
             class TestingRequest(Request):
                 """Test-only Request Class.
 
-                A testing request to use that will return a TestingFileStorage
+                A testing request to use that will return a MockingFileStorage
                 to test the uploading.
                 """
                 @property
@@ -136,7 +136,7 @@ class TestUploads(unittest.TestCase):
                     f = open(os.path.join('tests', 'fixtures', filename), 'rb')
                     print(os.path.join('tests', 'fixtures', filename))
                     print(type(f))
-                    d['image'] = TestingFileStorage(stream=BytesIO(f.read()),
+                    d['image'] = MockingFileStorage(stream=BytesIO(f.read()),
                                                     filename=filename)
                     f.close()
                     return d

--- a/webcompat/__init__.py
+++ b/webcompat/__init__.py
@@ -36,11 +36,11 @@ import webcompat.views  # noqa
 from webcompat import webhooks  # noqa
 
 # register blueprints
-from webcompat.api.endpoints import api  # noqa
+from webcompat.api.endpoints import api_bp  # noqa
 from webcompat.api.uploads import uploads  # noqa
 from webcompat.error_handlers import error_handlers  # noqa
 
-for blueprint in [api, error_handlers, uploads]:
+for blueprint in [api_bp, error_handlers, uploads]:
     app.register_blueprint(blueprint)
 
 

--- a/webcompat/api/endpoints.py
+++ b/webcompat/api/endpoints.py
@@ -27,13 +27,13 @@ from webcompat.helpers import normalize_api_params
 from webcompat.helpers import proxy_request
 from webcompat import limiter
 
-api = Blueprint('api', __name__, url_prefix='/api')
+api_bp = Blueprint('api_bp', __name__, url_prefix='/api')
 JSON_MIME_HTML = 'application/vnd.github.v3.html+json'
 ISSUES_PATH = app.config['ISSUES_REPO_URI']
 REPO_PATH = ISSUES_PATH[:-7]
 
 
-@api.route('/issues/<int:number>')
+@api_bp.route('/issues/<int:number>')
 @mockable_response
 def proxy_issue(number):
     """XHR endpoint to get issue data from GitHub.
@@ -44,7 +44,7 @@ def proxy_issue(number):
     return api_request('get', path, mime_type=JSON_MIME_HTML)
 
 
-@api.route('/issues/<int:number>/edit', methods=['PATCH'])
+@api_bp.route('/issues/<int:number>/edit', methods=['PATCH'])
 @mockable_response
 def edit_issue(number):
     """XHR endpoint to push back edits to GitHub for a single issue.
@@ -70,7 +70,7 @@ def edit_issue(number):
     abort(403)
 
 
-@api.route('/issues')
+@api_bp.route('/issues')
 @mockable_response
 def proxy_issues():
     """List all issues from GitHub on the API endpoint."""
@@ -89,7 +89,7 @@ def proxy_issues():
     return api_request('get', path, params=params)
 
 
-@api.route('/issues/<username>/<parameter>')
+@api_bp.route('/issues/<username>/<parameter>')
 @mockable_response
 def get_user_activity_issues(username, parameter):
     """Return issues related to a user at the API endpoint.
@@ -116,7 +116,7 @@ def get_user_activity_issues(username, parameter):
     return api_request('get', path, params=params)
 
 
-@api.route('/issues/category/<issue_category>')
+@api_bp.route('/issues/category/<issue_category>')
 @mockable_response
 def get_issue_category(issue_category):
     """Return all issues for a specific category."""
@@ -135,7 +135,7 @@ def get_issue_category(issue_category):
         abort(404)
 
 
-@api.route('/issues/search')
+@api_bp.route('/issues/search')
 @mockable_response
 @limiter.limit('30/minute',
                key_func=lambda: session.get('username', 'proxy-user'))
@@ -169,7 +169,7 @@ def get_search_results(query_string=None, params=None):
                        mime_type=JSON_MIME_HTML)
 
 
-@api.route('/issues/<int:number>/comments', methods=['GET', 'POST'])
+@api_bp.route('/issues/<int:number>/comments', methods=['GET', 'POST'])
 @mockable_response
 def proxy_comments(number):
     """XHR endpoint to get issues comments from GitHub.
@@ -188,7 +188,7 @@ def proxy_comments(number):
                            mime_type=JSON_MIME_HTML)
 
 
-@api.route('/issues/<int:number>/labels', methods=['POST'])
+@api_bp.route('/issues/<int:number>/labels', methods=['POST'])
 def modify_labels(number):
     """XHR endpoint to modify issue labels.
 
@@ -205,7 +205,7 @@ def modify_labels(number):
         abort(403)
 
 
-@api.route('/issues/labels')
+@api_bp.route('/issues/labels')
 @mockable_response
 def get_repo_labels():
     """XHR endpoint to get all possible labels in a repo."""

--- a/webcompat/templates/form.html
+++ b/webcompat/templates/form.html
@@ -121,7 +121,7 @@
       </div>
       {{ form.contact(class_='form-field text-field',
          placeholder='GitHub nickname',
-         pattern='^[a-z\d](?:[a-z\d]|-(?=[a-z\d])){0,38}$') }}
+         pattern='^[a-z\\d](?:[a-z\\d]|-(?=[a-z\\d])){0,38}$') }}
          <span class="form-input-validation check error js-bug-form-contact"></span>
     </div>
     {% endif %}

